### PR TITLE
Minimal Varianсe Sampling (MVS) booster

### LINKED ISSUE #2644: Minimal Variance Sampling in Stochastic Gradient Boosting
## Summary

MVS can be considered as an improved version of the Gradient-based One-Side Sampling (GOSS, see details in the paper) implemented in LightGBM, which samples a given number of top examples by values |gi| with the probability 1 and samples other examples with the same fixed probability. Due to the theoretical basis, MVS provides a lower variance of the estimate Eg than GOSS.

### Docs:
- https://catboost.ai/docs/concepts/algorithm-main-stages_bootstrap-options.html
### Code:
- https://github.com/catboost/catboost/commit/0bd529cbee49408ee9607885c9170bc037a0e2af
- https://github.com/catboost/catboost/pull/1005/files
### NeurIPS 2019 Poster:
- https://github.com/ibr11/mvs_paper/raw/master/mvs_poster.pdf
### NeurIPS 2019 Paper:
- http://papers.nips.cc/paper/9645-minimal-variance-sampling-in-stochastic-gradient-boosting.pdf

### PR DESCRIPTION
**Pull request by StrikerRUS**
Original PR #4266.
Fixed #2644.

Opening this PR to not accidentally remove underlying branch or forget about it existence.

### CODE REVIEW

File: include/LightGBM/config.h [hunk 1]
```diff
@@ -158,6 +158,7 @@ struct Config {
   // desc = ``rf``, Random Forest, aliases: ``random_forest``
   // desc = ``dart``, `Dropouts meet Multiple Additive Regression Trees <https://arxiv.org/abs/1505.01866>`__
   // descl2 = **Note**: internally, LightGBM uses ``gbdt`` mode for the first ``1 / learning_rate`` iterations
+  // desc = ``mvs``, Minimal variance sampling <https://arxiv.org/abs/1910.13204>__
```
Comment (jameslamb) [inline include/LightGBM/config.h:161]:
```suggestion
  // desc = ``mvs``, `Minimal variance sampling <https://arxiv.org/abs/1910.13204>`__
```

I think these extra backticks are necessary for this to format correctly as a hyperlink.

See

https://github.com/microsoft/LightGBM/blob/61e464bc02bd325672ce71daabfa57d569cf02bc/include/LightGBM/config.h#L159

and

https://sublime-and-sphinx-guide.readthedocs.io/en/latest/references.html#links-to-external-web-pages

File: include/LightGBM/config.h [hunk 2]
```diff
@@ -329,6 +330,25 @@ struct Config {
...
+
+  // default = false
+  // desc = use adaptive variant of mvs boosting
+  // desc = used only in ``mvs``
+  bool mvs_adaptive = false;
+
+  // default = 256000
+  // check = >0
+  // desc = used in MVS boosting training. If dataset size is greater than ``mvs_max_sequential_size``, then threshold
+  // desc = for MVS is chosen for each thread independently.
```
Comment (jameslamb) [inline include/LightGBM/config.h:347]:
```suggestion
  // desc = used in MVS boosting training. If dataset size is greater than ``mvs_max_sequential_size``, then threshold for MVS is chosen for each thread independently.
```

each `// desc =` line becomes 1 bullet point at https://lightgbm.readthedocs.io/en/latest/Parameters.html, so single sentences shouldn't be broken onto multiple lines this way

File: tests/cpp_tests/test_mvs_threshold_search.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
+/*!
+ * Copyright (c) 2021 Microsoft Corporation. All rights reserved.
```
Comment (jameslamb) [inline tests/cpp_tests/test_mvs_threshold_search.cpp:2]:
```suggestion
 * Copyright (c) 2023 Microsoft Corporation. All rights reserved.
```

I think this should be 2023, since this will be a new file added this year.

File: src/boosting/mvs.hpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
+/*!
+ * Copyright (c) 2017 Microsoft Corporation. All rights reserved.
```
Comment (jameslamb) [inline src/boosting/mvs.hpp:2]:
```suggestion
 * Copyright (c) 2023 Microsoft Corporation. All rights reserved.
```

File: src/boosting/mvs.cpp [hunk 5]
```diff
@@ -0,0 +1,180 @@
+/*!
+ * Copyright (c) 2021 Microsoft Corporation. All rights reserved.
```
Comment (jameslamb) [inline src/boosting/mvs.cpp:2]:
```suggestion
 * Copyright (c) 2023 Microsoft Corporation. All rights reserved.
```

File: include/LightGBM/config.h [hunk 6]
```diff
@@ -329,6 +330,25 @@ struct Config {
...
+  // desc = used only in ``mvs``
+  // desc = **Note**: on small dataset setting this parameter less than size of dataset may produce results depending on number of threads
```
Comment (jameslamb) [inline include/LightGBM/config.h:349]:
```suggestion
  // desc = **Note**: on small dataset setting this parameter less than size of dataset may produce non-deterministic results depending on number of threads
```

I'm not sure what this statement was supposed to be, but I'm fairly sure it wasn't just "may produce results". Is this not about the results being non-deterministic? Or should it maybe be "bad" results?

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks for reviving this @shiyu1994 ! I left a few minor comments for your consideration, but I'm not qualified to review this thoroughly. Hopefully @guolinke will be able to review.

Please `@` me if you need help with the failing CI jobs.

File: include/LightGBM/utils/array_args.h [hunk 7]
```diff
@@ -185,6 +186,36 @@ class ArrayArgs {
...
+        current_sum_small += sum_small + sum_middle;
+        end = middle_begin;
+      } else {
+        big_grad_size += n_middle + large_size;
+        begin = middle_end;
+      }
+    }
+
+    return current_sum_small / (sample_size - big_grad_size + kEpsilon);
+  }
```
Comment (guolinke) [inline include/LightGBM/utils/array_args.h:218]:
```suggestion
  static double CalculateThresholdMVS(std::vector<VAL_T>* gradients, data_size_t begin, data_size_t end,
                                             const double sample_size) {
    auto& ref_gradients = *gradients;
    double current_sum_small = 0.0;
    data_size_t big_grad_size = 0;

    while (begin != end) {
      data_size_t middle_begin = begin - 1, middle_end = end;
      ArrayArgs<score_t>::Partition(gradients, begin, end, &middle_begin, &middle_end);
      ++middle_begin;  // for half intervals
      const data_size_t n_middle = middle_end - middle_begin;
      const data_size_t large_size = middle_begin - begin;

      const double sum_small = std::accumulate(gradients->begin() + middle_end, gradients->begin() + end, 0.0);
      const double sum_middle = ref_gradients[middle_begin] * n_middle;

      const double
          current_sampling_rate = (current_sum_small + sum_small) / ref_gradients[middle_begin] + big_grad_size + n_middle + large_size;

      if (current_sampling_rate > sample_size) {
        current_sum_small += sum_small + sum_middle;
        end = middle_begin;
      } else {
        big_grad_size += n_middle + large_size;
        begin = middle_end;
      }
    }

    return current_sum_small / (sample_size - big_grad_size + kEpsilon);
  }
```

**Review by guolinke — APPROVED** [review]
Thank you, LGTM except a comment.

**Comment by shiyu1994** [comment]
This is still in progress, and I need to add some test cases for MVS before merging this.

File: include/LightGBM/utils/array_args.h (hunk 7 above)
Comment (shiyu1994) [reply include/LightGBM/utils/array_args.h:218]: It seems to be a unchanged change suggestion?

Comment (shiyu1994) [reply include/LightGBM/utils/array_args.h:218]: Sorry for missing the creating of reference variable.

File: include/LightGBM/config.h (hunk 6 above)
Comment (shiyu1994) [reply include/LightGBM/config.h:348]: I think it should be non-deterministic. Thanks for the fix.

File: src/boosting/goss.hpp [hunk 8]
```diff
@@ -84,20 +85,23 @@ class GOSSStrategy : public SampleStrategy {
...
     CHECK_LE(config_->top_rate + config_->other_rate, 1.0f);
     CHECK(config_->top_rate > 0.0f && config_->other_rate > 0.0f);
-    if (config_->bagging_freq > 0 && config_->bagging_fraction != 1.0f) {
-      Log::Fatal("Cannot use bagging in GOSS");
-    }
+
     Log::Info("Using GOSS");
     balanced_bagging_ = false;
     bag_data_indices_.resize(num_data_);
+    #ifdef USE_CUDA_EXP
```
Comment (jameslamb) [inline src/boosting/goss.hpp:92]: I'm surprised to see new commits referencing `cuda_exp`. `cuda_exp` was completely removed in #5677.

Please update this branch to latest `master` and replace any `cuda_exp` references with `cuda`.

**Review by jameslamb — CHANGES_REQUESTED** [review]

**Comment by jameslamb** [comment]
It's been over a year since the last commit to this PR, so I'm assuming that you're no longer working on this @shiyu1994. I'm closing this due to lack of activity, to make that clear to anyone else coming to the project interested in working on #2644.
